pkg/reconciler/workspaceshard/controller.go

```diff
@@ -0,0 +1,298 @@
+	if err != nil {
+		klog.Errorf("invalid key: %q: %v", key, err)
+		return nil
+	}
+	if namespace != "" {
+		klog.Errorf("namespace %q found in key for cluster-wide WorkspaceShard object", namespace)
+		return nil
+	}
+	clusterName, name := clusters.SplitClusterAwareKey(clusterAwareName)
+
+	obj, err := c.workspaceShardLister.Get(key) // TODO: clients need a way to scope down the lister per-cluster
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return nil // object deleted before we handled it
+		}
+		return err
+	}
+	previous := obj
+	obj = obj.DeepCopy()
+
+	if err := c.reconcile(ctx, obj); err != nil {
+		return err
+	}
+
+	// If the object being reconciled changed as a result, update it.
+	if !equality.Semantic.DeepEqual(previous.Status, obj.Status) {
+		oldData, err := json.Marshal(tenancyv1alpha1.WorkspaceShard{
+			Status: previous.Status,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to Marshal old data for workspace shard %q|%q/%q: %w", clusterName, namespace, name, err)
+		}
+
+		newData, err := json.Marshal(tenancyv1alpha1.WorkspaceShard{
+			ObjectMeta: metav1.ObjectMeta{
+				UID:             previous.UID,
+				ResourceVersion: previous.ResourceVersion,
+			}, // to ensure they appear in the patch as preconditions
+			Status: obj.Status,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to Marshal new data for workspace shard %q|%q/%q: %w", clusterName, namespace, name, err)
+		}
+
+		patchBytes, err := jsonpatch.CreateMergePatch(oldData, newData)
+		if err != nil {
+			return fmt.Errorf("failed to create patch for workspace shard %q|%q/%q: %w", clusterName, namespace, name, err)
+		}
+		_, uerr := c.kcpClient.Cluster(clusterName).TenancyV1alpha1().WorkspaceShards().Patch(ctx, obj.Name, types.MergePatchType, patchBytes, metav1.PatchOptions{}, "status")
+		return uerr
+	}
+
+	return nil
+}
+
+func (c *Controller) reconcile(ctx context.Context, workspaceShard *tenancyv1alpha1.WorkspaceShard) error {
+	secret, err := c.secretLister.Secrets(workspaceShard.Spec.Credentials.Namespace).Get(clusters.ToClusterAwareKey(workspaceShard.ClusterName, workspaceShard.Spec.Credentials.Name))
+	if errors.IsNotFound(err) {
+		conditions.MarkFalse(workspaceShard, tenancyv1alpha1.WorkspaceShardCredentialsValid, tenancyv1alpha1.WorkspaceShardCredentialsReasonMissing, "Referenced secret %s/%s could not be found.", workspaceShard.Spec.Credentials.Namespace, workspaceShard.Spec.Credentials.Name)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	data, ok := secret.Data[tenancyv1alpha1.WorkspaceShardCredentialsKey]
+	if !ok {
+		conditions.MarkFalse(workspaceShard, tenancyv1alpha1.WorkspaceShardCredentialsValid, tenancyv1alpha1.WorkspaceShardCredentialsReasonInvalid, "Referenced secret %s/%s did not contain key %s.", workspaceShard.Spec.Credentials.Namespace, workspaceShard.Spec.Credentials.Name, tenancyv1alpha1.WorkspaceShardCredentialsKey)
+		return nil
+	}
+
+	cfg, err := clientcmd.RESTConfigFromKubeConfig(data)
```

Comment:
Is this correct? Should I just join the host+apipath here? Is there anything else?